Improve editor support for icons of custom, scripted, and GDExtension classes

YuriSizov opened the pull request:
This is three PRs in a trench coat! But it all makes sense, let me explain.

### Streamline class icon resolution in the editor

The first commit reworks some disjointed operations related to class icon lookup. It tries to remove a bunch of duplicated logic from `EditorNode` and various other editor classes.

From my search, most of the editor uses `EditorNode::get_object_icon` and `EditorNode::get_class_icon`, but some important places weren't. Worse yet, those methods, while introduced at the same time as a part of #21717, worked slightly differently. They followed mostly the same base logic, but had discrepancies for no good reason. So I unified their internal logic (and moved it alongside a cache map to `EditorData` for good measure), leaving them only as interfaces for two types of request: by object reference, and by name.

I've also made sure that these methods are used in such places as Inspector (for categories) and Editor Help (for the header and inheritance chains). I also added the necessary call to `EditorQuickOpen`, because existing logic was faulty. It tried to fetch an icon from the editor theme based on a script name, but that would never work (implemented in #62417). I'm worried a bit that this may become much slower than before, but there are a bunch of caches in place, so hopefully it scales well for bigger projects.

I removed the meta hack for icons, because there doesn't seem to be anything that uses it in the editor anymore. It was introduced sometime around #20560, #21717 refactored it a bunch, and #30697 removed it even more. So currently it's only read in our codebase, but never set. Something may still rely on this existing, as a hack, but I don't think it's worth to support such undocumented "feature".

### Make icons of scripted and custom classes fit the editor UI

The second commit builds on top of the first one, and tackles the long-standing issue with custom types having icons overflow various parts of the editor.

First, I made the issue worse. I noticed that icons loaded as a part of the logic modified in commit 1 would never respect editor-specific import options, and thus would never be auto-scaled and auto-converted to match editor colors. So I switched `ImageLoader` for `ResourceLoader`, but that made every one of these icons to blow out in the same way custom type icons blow out. So it was a great opportunity to work on that as well.

I've incorporated (and superseded) #71817 and #71818, improving upon them. There were more issues, however, in the scene tree, in the create dialog, in documentation. Several of these cases could be patched with the existing API that we have for the corresponding controls. There were a couple of exceptions exceptions: `PopupMenu`, `TabBar`, and `Button`.

For the `PopupMenu` and `TabBar` the solution was simple, as they are pretty close visually to `Tree` and have similar sizing considerations. So I borrowed the approach from `Tree` and `TreeItem` here. Now items in `PopupMenu` and tabs in `TabBar` can set the upper limit to the icon's width with `PopupMenu::set_item_icon_max_width` and `TabBar::set_tab_icon_max_width`, respectively. The height is scaled according to the icon's own size ratio.

For `Button` there was already a handy `expand_icon` property, but it resulted in issues when the button was way bigger than the icon. So on top of that I had to introduce a limiter for the width as well. I used a theme constant here, `icon_max_width`, so that the configuration can be shared and applied globally. This is similar to #71872, but only for the width, so it works exactly like `Tree`, `PopupMenu`, and `TabBar`.

In fact, I also added `icon_max_width` theme constant to `Tree`, `PopupMenu`, and `TabBar`, so these controls can also have shared configuration within a project. This limit is applied before the max width set on individual items, if any. To help us organize this further I added a new editor theme constant for shared use, `class_icon_size`. This was a suggestion from @akien-mga in one of the linked PRs, and I agreed with the idea.

![godot windows editor dev x86_64_2023-03-29_20-40-10](https://user-images.githubusercontent.com/11782833/228643790-d34289f4-c5fd-4870-b0d0-a2c9d86ad62e.png)

### Add support for icons in GDExtension classes

Finally, the third commit enhances the existing icons feature to add support for GDExtension.

This is based on @akien-mga's work (see https://github.com/akien-mga/godot/commit/242ce09cc33e8085dbeebe5f1cd03907660c3d1f), I didn't change much from that. Only adjusted the code to match my rework, and added a clean-up step to `GDExtensionManager::unload_extension` missing from the original. That's it.

Icons can be added to GDExtension classes in a way proposed by @reduz on RC. Just add the following to your `*.gdextension` file:

```
[icons]
Threen = "res://bin/icon_tween.svg"
```

Paths are in the project scope, same as paths for binaries. On the left side is the name of a class, on the right the path to its icon. Icons aren't bundled in DLLs, as this would complicate the build setup, as far as I understand. If GDExtenstion API is not so great here, we can rework it later. The important part is the editor integration, the behind-the-scenes can be reworked as needed, if another consensus is reached.

![godot windows editor dev x86_64_2023-03-29_20-40-32](https://user-images.githubusercontent.com/11782833/228643863-1d4ad2e0-3c10-460a-9cc5-897a74d9f024.png)

-----

So here we are, a bunch of fixes and improvements, that go well together. If we're okay with the new features, these commits could be cherry-picked in 4.0.x, I think. At least the first two.

Fixes #57286.
Fixes #68962.
Closes #69212.
Closes https://github.com/godotengine/godot-cpp/issues/863.

bruvzg commented on scene/gui/popup_menu.cpp:
```diff
@@ -62,6 +62,11 @@ Size2 PopupMenu::_get_contents_minimum_size() const {
 		const_cast<PopupMenu *>(this)->_shape_item(i);
 
 		Size2 icon_size = items[i].get_icon_size();
+		if (items[i].icon_max_width > 0 && icon_size.width > items[i].icon_max_width) {
+			icon_size.height = icon_size.height * items[i].icon_max_width / icon_size.width;
+			icon_size.width = items[i].icon_max_width;
+		}
```
The same should be applied to global menu icons here (it's using child `PopupMenu`s to build global menu):
https://github.com/godotengine/godot/blob/c29866dbc05694594b9be68b8d451816c7a0889d/scene/gui/menu_bar.cpp#L231

Something like:
```c++
Ref<Texture2D> icon = p_child->get_item_icon(i);
if (icon.is_valid()) {
	int icon_max_width = p_child->get_item_icon_max_width(i);
	Size2i icon_size = icon->get_size();
	if (icon_max_width > 0 && icon_size.width > icon_max_width) {
		icon_size.height = icon_size.height * icon_max_width / icon_size.width;
		icon_size.width = icon_max_width;
		icon = icon->duplicate();
		icon->resize(icon_size);
	}
}
DisplayServer::get_singleton()->global_menu_set_item_icon(p_menu_name, index, icon);
```
or maybe `item_icon_max_width` can be added to the global menu API.

YuriSizov replied:
Either way works, though I'm not a fan of duplicating the resource. But is it actually possible to overflow the size of a native menu? I'd think they have a hard limit and would scale the icon already.

YuriSizov replied:
We also don't have a resize method on `Texture2D`, we'd need to create an `ImageTexture` from the image, or something like that. So I'm still wondering if it's actually possible to overflow native menus. I don't have a Mac to test it myself.

bruvzg replied:
It is possible to use any size of the image in the macOS menu, and it will stretch the menu, but there's already hard-coded `resize(16, 16, Image::INTERPOLATE_LANCZOS);` in the macOS DisplayServer. So it should be good as is (probably `item_icon_max_width` should replace this hardcoded size in the future, just in case someone wants bigger menu icons).

<img width="588" alt="Screenshot 2023-03-30 at 14 24 32" src="https://user-images.githubusercontent.com/7645683/228821361-ed2e4891-a744-4b8d-9db4-976f8bea1489.png">

akien-mga reviewed (approved):
Looks really good!

YuriSizov commented on core/extension/gdextension_manager.cpp:
```diff
@@ -50,6 +50,11 @@ GDExtensionManager::LoadStatus GDExtensionManager::load_extension(const String &
 	}
+
+	for (const KeyValue<String, String> &kv : extension->class_icon_paths) {
+		gdextension_class_icon_paths[kv.key] = kv.value;
```
Could possibly check in `gdextension.cpp` when we fill its internal map that keys match registered types? Two extensions shouldn't be able to register a class with the same name, so this would be sufficient, wdyt?

bruvzg replied:
It should not be able to register class with the same name, but can have icons for anything in the `.gdextension`, so I maybe it should only allow icons for the registered classes (but this is probably not trivial, since classes can be registered on different init levels).

YuriSizov replied:
Yeah, I would just not support arbitrary icons, at least for now. I can see that checking against registered classes is indeed not trivial, though. Will need to look more into it, but I guess we can address it in a follow-up?

akien-mga replied:
For the time being I think it's ok if we just let whichever gets registered first or last take precedence. Naming conflicts should be rare and we can handle this later on.

paddy-exe commented:
> Icons can be added to GDExtension classes in a way proposed by @reduz on RC. Just add the following to your config file
Just to clarify: By config file you mean the `.gdextension` or an actual config file?

YuriSizov commented:
@paddy-exe I mean the `.gdextension`, yeah. Let me edit to clarify.

KoBeWi commented on core/extension/gdextension_manager.h:
```diff
@@ -59,6 +60,9 @@ class GDExtensionManager : public Object {
 	Ref<GDExtension> get_extension(const String &p_path);
 
+	bool class_has_icon_path(const String &p_class) const;
+	String class_get_icon_path(const String &p_class) const;
```
```suggestion
	const String &class_get_icon_path(const String &p_class) const;
```
I think this should be possible.

YuriSizov replied:
Didn't work in the end, because of `return ""` creating a local/temporary variable, and I'm not confident how to solve this.

KoBeWi commented on core/extension/gdextension_manager.cpp:
```diff
@@ -95,6 +105,17 @@ Ref<GDExtension> GDExtensionManager::get_extension(const String &p_path) {
 	return E->value;
 }
 
+bool GDExtensionManager::class_has_icon_path(const String &p_class) const {
```
This is basically equivalent of `!class_get_icon_path().is_empty()`.

YuriSizov replied:
Yeah, but it's explicit this way, so I don't mind.

KoBeWi commented:
Tested custom type icons (added via `add_custom_type()`). The icon appears correctly in new node dialog
![image](https://user-images.githubusercontent.com/2223172/228684398-ba408dd2-95c8-404c-bb67-dca2050deb3c.png)
Except in recent nodes
![image](https://user-images.githubusercontent.com/2223172/228684423-3f5dfd7e-d02b-40cb-824a-d60898ed87ac.png)
Does not appear in scene tree and on scene tabs
![image](https://user-images.githubusercontent.com/2223172/228684472-f79ed042-3f65-475a-81e3-b2a7459636a1.png)

KoBeWi commented on scene/gui/popup_menu.h:
```diff
@@ -222,6 +223,7 @@ class PopupMenu : public Popup {
 	void set_item_text_direction(int p_idx, Control::TextDirection p_text_direction);
 	void set_item_language(int p_idx, const String &p_language);
 	void set_item_icon(int p_idx, const Ref<Texture2D> &p_icon);
+	void set_item_icon_max_width(int p_idx, int p_width);
```
I think it would be great if this was exposed (like it's already in TreeItem).
I needed it at least once in my projects.

KoBeWi commented on editor/editor_data.cpp:
```diff
@@ -1028,8 +1031,73 @@ void EditorData::script_class_load_icon_paths() {
+}
+
+Ref<Texture2D> EditorData::_load_script_icon(const String &p_path) const {
+	if (p_path.length() && ResourceLoader::exists(p_path)) {
```
```suggestion
	if (!p_path.is_empty() && ResourceLoader::exists(p_path)) {
```

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -4480,106 +4467,80 @@ void EditorNode::_pick_main_scene_custom_action(const String &p_custom_action_na
-Ref<Texture2D> EditorNode::get_class_icon(const String &p_class, const String &p_fallback) const {
-	ERR_FAIL_COND_V_MSG(p_class.is_empty(), nullptr, "Class name cannot be empty.");
+Ref<Texture2D> EditorNode::get_object_icon(const Object *p_object, const String &p_fallback) {
+	ERR_FAIL_COND_V_MSG(!p_object, nullptr, "Object cannot be null.");
```
```suggestion
	ERR_FAIL_NULL(p_object, nullptr, "Object cannot be null.");
```

YuriSizov replied:
`ERR_FAIL_NULL_V_MSG` doesn't make it much shorter, but I guess it's easier to read :)

KoBeWi replied:
It has a different error message.

KoBeWi commented:
Also this broke
![image](https://user-images.githubusercontent.com/2223172/228690027-4d2ab3d7-f531-4a56-a412-d5a51190f613.png)

YuriSizov commented:
Ah, I guess we still need some form of width limiter to button icons. I'll add one via the theme, similar to https://github.com/godotengine/godot/pull/71872, but only for width and working exactly like in Tree and PopupMenu.

KoBeWi commented:
Scene icons still need fixing.

YuriSizov commented:
Okay, I fixed all the reported issues regarding the fetching and sizing of the icons. And a couple of unreported ones. Compared to the original submission this PR now includes:

- `TabBar` receiving a similar treatment to `PopupMenu`;
- `Button` having a new theme constant to apply an upper limit to the icon's size;
- `Tree`, `TabBar`, and `PopupMenu` having a similar new theme constant, which is used before the item's/tab's individual limit is applied;
- New methods exposed on `PopupMenu` and `TabBar` to set up the per-item limit from scripts, docs are added and adjusted where relevant;
- Icon lookup logic is improved more, and should now cover correctly all three types of custom classes across the UI.

I still haven't added anything to `MenuBar` (see discussion on that suggestion), neither have I added any sanity checks to GDExtension possible icon conflicts. I can look more into it now, or I can look into it later, in a follow-up (or somebody else can).

YuriSizov commented:
@KoBeWi I think it should be good to retest now. See the changes listed above. And here's a custom class in various poses.

![godot windows editor dev x86_64_2023-03-30_14-53-06](https://user-images.githubusercontent.com/11782833/228846976-32d7cec4-59f6-4abd-b4ce-0b08731a0576.png)

KoBeWi commented on scene/gui/popup_menu.cpp:
```diff
@@ -47,6 +47,26 @@ String PopupMenu::_get_accel_text(const Item &p_item) const {
+	if (theme_cache.icon_max_width > 0) {
+		max_width = theme_cache.icon_max_width;
+	}
+	if (item.icon_max_width > 0 && item.icon_max_width < max_width) {
```
If `theme_cache.icon_max_width` is 0, `item.icon_max_width` has no effect.

KoBeWi commented on scene/gui/tab_bar.cpp:
```diff
@@ -1305,6 +1331,27 @@ int TabBar::get_tab_width(int p_idx) const {
+	if (theme_cache.icon_max_width > 0) {
+		icon_max_width = theme_cache.icon_max_width;
+	}
+	if (tab.icon_max_width > 0 && tab.icon_max_width < icon_max_width) {
```
Same here.

YuriSizov replied:
Yes, naturally, I copied the same code in a few places. I'll address it everywhere.

YuriSizov replied:
Yes, naturally, I copied the same code in a few places. I'll address it everywhere.

KoBeWi commented on scene/gui/tree.cpp:
```diff
@@ -1654,6 +1652,25 @@ void Tree::_update_theme_item_cache() {
+	if (theme_cache.icon_max_width > 0) {
+		max_width = theme_cache.icon_max_width;
+	}
+	if (p_cell.icon_max_w > 0 && p_cell.icon_max_w < max_width) {
```
I didn't test this one, but likely same.

YuriSizov replied:
I figured you were just reading the code, no worries :)

YuriSizov committed:
Make icons of scripted and custom classes fit the editor UI

Also:
- Add an option to limit the icon size in PopupMenu.
This is similar to how this works in Tree and TreeItem.
- Add the same option to TabBar.
- Add a theme constant for Tree, PopupMenu, Button, and
TabBar to apply this limit on the control level.

Co-authored-by: Daylily-Zeleen <daylily-zeleen@foxmail.com>

YuriSizov committed:
Streamline class icon resolution in the editor

YuriSizov committed:
Add support for icons in GDExtension classes

Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

YuriSizov commented:
Fixed the checks, and also fixed the tabs drag preview not respecting the max size of the icon. I also rebased for good measure, so the force-push diff is not going to be helpful, sorry.

akien-mga commented:
Thanks!
